cli/nodeutil.ts

```diff
@@ -489,43 +489,76 @@ export function fileExistsSync(p: string): boolean {
 export let lastResolveMdDirs: string[] = []
 
 // returns undefined if not found
-export function resolveMd(root: string, pathname: string): string {
-
+export function resolveMd(root: string, pathname: string, md?: string): string {
     const docs = path.join(root, "docs");
 
-    let tryRead = (fn: string) => {
+    const tryRead = (fn: string) => {
         if (fileExistsSync(fn + ".md"))
             return fs.readFileSync(fn + ".md", "utf8")
         if (fileExistsSync(fn + "/index.md"))
             return fs.readFileSync(fn + "/index.md", "utf8")
         return null
     }
 
-    let targetMd = tryRead(path.join(docs, pathname))
+    const targetMd = md ? md : tryRead(path.join(docs, pathname))
     if (targetMd && !/^\s*#+\s+@extends/m.test(targetMd))
         return targetMd
 
-    let dirs = [
+    const dirs = [
         path.join(root, "/node_modules/pxt-core/common-docs"),
-    ]
-    lastResolveMdDirs = dirs
-    for (let pkg of pxt.appTarget.bundleddirs) {
-        let d = path.join(pkg, "docs");
-        if (!path.isAbsolute(d)) d = path.join(root, d);
-        dirs.push(d)
-
-        let cfg = readPkgConfig(path.join(d, ".."))
-        for (let add of cfg.additionalFilePaths)
-            dirs.push(path.join(d, "..", add, "docs"))
-    }
-    for (let d of dirs) {
-        let template = tryRead(path.join(d, pathname))
+        ...getBundledPackages()
+    ];
+
+    for (const d of dirs) {
+        const template = tryRead(path.join(d, pathname))
         if (template)
             return pxt.docs.augmentDocs(template, targetMd)
     }
     return undefined;
 }
 
+export function getBundledPackages(): string[] {
+    const handledDirectories = {};
+    const docFolders: string[] = [];
+
+    for (const bundledDir of pxt.appTarget.bundleddirs || []) {
+        getPackageDocs(bundledDir, docFolders, handledDirectories);
+    }
+
+    return docFolders;
+
+    function getPackageDocs(dir: string, folders: string[], resolvedDirs: pxt.Map<boolean>) {
+        if (resolvedDirs[dir])
+            return;
+
+        resolvedDirs[dir] = true;
+
+        const jsonDir = path.join(dir, "pxt.json");
+        const pxtjson = fs.existsSync(jsonDir) && (readJson(jsonDir) as pxt.PackageConfig);
+
+        if (pxtjson) {
+            if (pxtjson.additionalFilePath) {
+                getPackageDocs(path.join(dir, pxtjson.additionalFilePath), folders, resolvedDirs);
```

Comment:
I think this needs to be recursive - the pxt.json in additionalFilePath can have additionalFilePath, eg mixer---samd in Arcade would point to mixer---samd in common, which would point to mixer in common which has docs

Reply:
though I guess you won't see it now, because you're going over all bundled packages, which will include "mixer" in Arcade pointing to "mixer" in common. maybe that's enough.

Reply:
Ah, I am recursing here already - for each package, I'm starting a depth first traversal of each other package included as an `additionalFilePath` or in the `dependencies` to make sure all dependencies are included before the files that depend on them; I'll add some comments here as there are some invariants that need to be maintained to make sure everything gets outputted properly